app/web/src/features/communities/hooks.ts

```diff
@@ -36,21 +37,58 @@ import {
   useQueryClient,
   UseQueryOptions,
 } from "react-query";
+import { useHistory, useParams } from "react-router-dom";
+import { routeToCommunity } from "routes";
 import { service } from "service";
 
 export const useCommunity = (
-  id: number,
+  id: number | undefined,
   options?: Omit<
-    UseQueryOptions<Community.AsObject, GrpcError>,
-    "queryKey" | "queryFn"
+    UseQueryOptions<Community.AsObject | undefined, GrpcError>,
+    "queryKey" | "queryFn" | "enabled"
   >
-) =>
-  useQuery<Community.AsObject, GrpcError>(
-    communityKey(id),
-    () => service.communities.getCommunity(id),
-    options
+) => {
+  const {
+    communityId: communityIdFromUrl,
+    communitySlug: communitySlugFromUrl,
+  } = useParams<{
+    communityId?: string;
+    communitySlug?: string;
+  }>();
+
+  const communityId =
+    id ?? (communityIdFromUrl ? +communityIdFromUrl : undefined);
+
+  const queryResult = useQuery<Community.AsObject | undefined, GrpcError>(
+    communityKey(communityId || -1),
+    () =>
+      communityId
+        ? service.communities.getCommunity(communityId)
+        : Promise.resolve(undefined),
+    {
+      ...options,
+      enabled: !!communityId,
+    }
   );
 
+  const { data: { slug } = {} } = queryResult;
+
+  const history = useHistory();
+
+  useEffect(() => {
+    // guarantee the most recent slug is used if the community was loaded from url params
+    // if no slug was provided in the url, then also redirect to page with slug in url
+    if (!id && slug && slug !== communitySlugFromUrl) {
+      communityId && history.push(routeToCommunity(communityId, slug));
+    }
```

Comment:
```suggestion
    if (!id && slug && slug !== communitySlugFromUrl && communityId) {
      history.push(routeToCommunity(communityId, slug));
    }
```

Reply:
yeah I see what you mean here but it makes that if statement even less readable :(
I will change the code so we can infer the correct type for both slug and communityid. Since both of the variables can never be undefined after the query has been successfully fetched.